File: canwin.py
# O(n) memory and 0(n * k) time where k = arr.max

def can_win(arr):
    bool_arr = [False] * len(arr)
    for i in reversed(range(len(arr))):
        for j in reversed(range(arr[i])):
            k = j + 1
            if i + k >= len(arr) - 1 or bool_arr[i + k]:
                bool_arr[i] = True
                break
    return bool_arr[0]

# print(can_win([3, 0, 0, 1, 3]))


# O(1) memory and O(n) time

def can_win2(arr):
    farthest = 0
    last = len(arr) - 1
    i = 0
    while i <= farthest and farthest < last:
        farthest = max(farthest, arr[i] + i)
        i += 1
    return farthest >= last

print(can_win2([2, 0, 1, 3]))

File: arrays/canwin.js
 // O(n) memory and 0(n * k) time where k = arr.max

function canWin(array) {
  var boolArr = [];
  for (var i = 0; i < array.length; i++) {
    boolArr.push(false);
  }
  // const boolArr = new Array(array.length).fill(false);

  for (var i = array.length - 1; i >= 0; i--) {
    for (var j = array[i]; j > 0; j--) {
      if ((i + j) === array.length - 1 || boolArr[i + j] === true) {
        boolArr[i] = true;
        break;
      }
    }
  }

  return boolArr[0];
}
// console.log(canWin([2, 0, 0, 1, 3]));




// O(1) memory and O(n) time

function canWin2(array) {
  var farthest = 0;
  var last = array.length - 1;
  var i = 0;

  while (i <= farthest && farthest < last) {
    farthest = Math.max(farthest, array[i] + i);
    i++;
  }

  return farthest >= last;
}

console.log(canWin2([3, 0, 0, 1, 3]));
